lib/collections.go

```diff
@@ -948,6 +1020,44 @@ func zipLists(arg0, arg1 ref.Val) ref.Val {
 	return types.NewRefValMap(types.DefaultTypeAdapter, m)
 }
 
+func mapKeys(val ref.Val) ref.Val {
+	mapK, ok := val.(traits.Mapper)
+	if !ok {
+		return types.ValOrErr(mapK, "no such overload")
+	}
+	n, _ := mapK.Size().(types.Int)
+	new := make([]ref.Val, 0, n)
+	if mapK.Size() != types.IntZero {
+		m, err := mapK.ConvertToNative(refValMap)
+		if err != nil {
+			return types.NewErr("unable to convert map to native: %v", err)
+		}
+		for key, _ := range m.(map[ref.Val]ref.Val) {
+			new = append(new, key)
+		}
+	}
+	return types.NewRefValList(types.DefaultTypeAdapter, new)
+}
+
+func mapValues(val ref.Val) ref.Val {
```

Comment:
Similar comments as above, collect the keys, sort if possible and then collect the key-ordered values. There is an additional concern here, someone may use double keys (they probably shouldn't and we could be justified in not handling this particular case) and end up with a NaN key. To handle this, collect the values at the same time as the keys and cosort the values based on the keys.

Reply:
updated as suggested